functions/go/set-namespace/README.md

```diff
@@ -17,68 +17,114 @@ Namespaces are often used in the following scenarios:
+  If not, this function won't add new namespace.
+- This function updates `RoleBinding` and `ClusterRoleBinding` resources `subjects` element whose kind is `ServiceAccount`
+  and the subject's `namespace` is set and matches the "oldNamespace".
+- This function updates `CustomResourceDefinition` (CRD) `spec/conversion/webhook/clientConfig/service/namespace` field 
+  if the field is set and matches the "oldNamespace"
+- This function updates `APIService` `spec/service/namespace` field if the field is set and matches the "oldNamespace"
+- This function updates the KRM resources annotation `config.kubernetes.io/depends-on` if this annotation contains the 
+  matching namespace.
+
+### FunctionConfig
+
+This function supports the default `ConfigMap` as function config and a custom `SetNamespace`. See below examples
+
+ConfigMap as functionConfig
 ```yaml
 apiVersion: v1
-kind: ServiceAccount
-metadata:
-  name: sa
-  namespace: example
-  annotations:
-    config.kubernetes.io/depends-on: /namespaces/example/ServiceAccount/foo # <= this will NOT be updated (resource not declared)
----
-kind: RoleBinding
-apiVersion: rbac.authorization.k8s.io/v1
-metadata:
-  ...
-  annotations:
-    config.kubernetes.io/depends-on: /namespaces/example/ServiceAccount/sa # <== this will be updated (resource declared)
-subjects:
-  - kind: ServiceAccount
-    name: default # <================== name default is used
-    namespace: example # <================== this will be updated
-roleRef:
-  kind: Role
-  name: confluent-operator
-  apiGroup: rbac.authorization.k8s.io
+kind: ConfigMap
+data:
+  namespace: newNamespace # required
+  namespaceSelector: example # update namespace whose value is "example" to "newNamespace"
 ```
 
-This function can be used both declaratively and imperatively.
+SetNamespace as functionConfig
+```yaml
+apiVersion: fn.kpt.dev/v1alpha1
+kind: SetNamespace
+namespace: newNamespace # required
+namespaceSelector: example # update namespace whose value is "example" to "newNamespace"
+```
 
-### FunctionConfig
 
-There are 2 kinds of `functionConfig` supported by this function:
+### Three updating modes
 
-- `ConfigMap`
-- A custom resource of kind `SetNamespace`
+This function supports three modes to flexibly choose and update the target namespaces.
 
-To use a `ConfigMap` as the `functionConfig`, the desired namespace must be
-specified in the `data.namespace` field.
+##### Restrict Mode
+All target KRM resources namespace has to have the same value. All namespace will be updated to the new value. 
 
-To add a namespace `staging` to all resources, we use the
-following `functionConfig`:
+ConfigMap as functionConfig
+```yaml
+apiVersion: v1
+kind: ConfigMap
+data:
+  namespace: newNamespace # update all namespace fields to "newNamespace"
+```
+
+##### DefaultNamespace Mode
+
+The input `resourcelist.items` contains one and only one namespace object. The function matches the namespace `metadata.name`
```

Comment:
> contains one and only one namespace object

contains **one and only one** `Namespace` object

Reply:
done.